File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.py
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        
        fmap = {}
        for word in words:
            fmap[word] = fmap.get(word, 0)+1
        
        centerUsed = False
        result = 0
        
        for word in words:
            rev = word[::-1]
            
            if rev != word:
                if fmap[word]>0 and fmap.get(rev, 0)>0:
                    result += 4;
                    fmap[word] -= 1
                    fmap[rev] -= 1
            else:
                if fmap[word] >= 2:
                    result += 4
                    fmap[word] -= 2
                elif fmap[word] == 1 and centerUsed == False:
                    result += 2
                    fmap[word] -= 1
                    centerUsed = True
        return result
    
            
            

File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.java
class Solution {
    public int longestPalindrome(String[] words) {
        
        Map<String, Integer> fmap = new HashMap<>(); 
        for(String word: words){
            fmap.put(word, fmap.getOrDefault(word, 0) + 1);
        }
        
        int longestPalindromeWord = 0;
        boolean centerUsed = false;
        
        for(String word: words){
            
            String rev = new StringBuilder(word).reverse().toString();
            
            if(!rev.equals(word)){
                if(fmap.get(word)>0 && fmap.getOrDefault(rev, 0)>0){
                    longestPalindromeWord += 4;
                    fmap.put(word, fmap.get(word)-1);
                    fmap.put(rev, fmap.get(rev)-1);
                }
            }
            else{
                if(fmap.get(word) >= 2){
                    longestPalindromeWord += 4;
                    fmap.put(word, fmap.get(word)-2);
                }
                else if(fmap.get(word) ==1 && centerUsed == false){
                    longestPalindromeWord += 2;
                    fmap.put(word, fmap.get(word)-1);
                    centerUsed = true;
                }
            }
        }
        
        return longestPalindromeWord;
        
    }
}